metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -166,10 +166,56 @@ using CallOnes = std::function<void(const bit_vector &source_col,
                                     const uint64_t *pred_begin,
                                     const uint64_t *pred_end)>;
 
+void read_next_block(sdsl::int_vector_buffer<>::iterator *succ_it_p,
+                     sdsl::int_vector_buffer<1>::iterator *pred_boundary_it_p,
+                     sdsl::int_vector_buffer<>::iterator *pred_it_p,
+                     uint64_t block_size,
+                     std::array<std::vector<uint64_t>, 3> *out) {
+    auto &succ_it = *succ_it_p;
+    auto &pred_boundary_it = *pred_boundary_it_p;
+    auto &pred_it = *pred_it_p;
+
+    std::vector<uint64_t> &succ_chunk = out->at(0);
+    std::vector<uint64_t> &pred_chunk_idx = out->at(1);
+    std::vector<uint64_t> &pred_chunk = out->at(2);
+
+    #pragma omp parallel sections num_threads(2)
+    {
+        #pragma omp section
+        {
+            succ_chunk.resize(block_size);
+            for (uint64_t i = 0; i < block_size; ++i, ++succ_it) {
+                succ_chunk[i] = *succ_it;
+            }
+        }
+
+        #pragma omp section
+        {
+            // read predecessor offsets
+            pred_chunk_idx.resize(block_size + 1);
+            pred_chunk_idx[0] = 0;
+            for (uint64_t i = 1; i <= block_size; ++i) {
+                // find where the last predecessor for the node ends
+                pred_chunk_idx[i] = pred_chunk_idx[i - 1];
+                while (*pred_boundary_it == 0) {
+                    ++pred_chunk_idx[i];
+                    ++pred_boundary_it;
```

Comment:
what happens here if the file pointed to by `pred_boundary_it` is truncated? does calling `operator++` throw an exception if this tries to go beyond the end?

Reply:
It will pad with zeros, so this will be an infinite loop. Looks quite safe.

Reply:
I added an explicit check.